[RF Docs] Rails Internationalization (#57381)
rails/rails: guides/source/i18n.md

Reviewer: It feels like all the "Translations for *" sections deserve their own top section.
These are less about the I18n API, and more about build-in Rails functionality.
Optionally flatten the "Other I18n Helpers" section, to make them visible in the sidebar as well.

What do you think about something like:
- Overview of the I18n API Features
  - Looking up Translations
  - Pluralization
  - Setting and Passing a Locale
  - Using Safe HTML Translations
- Build-in Rails Translations (or maybe "Rails Translations")
  - Active Record Models
  - Action Mailer E-Mail Subjects
  - Form Helpers
  - Action View Helper Methods
  - Active Model Methods
  - Active Support Methods

Author: ✅

Reviewer: If the `divs` are removed this should be deleted (the image as well):
```suggestion
```

**DO NOT READ THIS FILE ON GITHUB, GUIDES ARE PUBLISHED ON <https://guides.rubyonrails.org>.**

Internationalization (I18n)
===========================

> The Ruby `i18n` gem, which ships with Ruby on Rails, provides an easy-to-use and extensible framework for translating your application into a single custom language or for providing multi-language support.

Reviewer 2: Could we link the gem?
```suggestion
The Ruby [`i18n` gem](https://github.com/ruby-i18n/i18n), which ships with Ruby on Rails, provides an easy-to-use and extensible framework for translating your application into a single custom language or for providing multi-language support.
```

Reviewer 2: Nevermind, I see we link below. Maybe this is unecessary. Or maybe we should move the link here?

After reading this guide, you will know:

* How I18n works in Ruby on Rails
* How to configure, set, and switch locales in a Rails application
* How to organize translations and use the I18n API in views, controllers, and models
* How to use I18n to translate Active Record errors or Action Mailer E-mail subjects
* How to customize backends and handle translation-related exceptions

--------------------------------------------------------------------------------
Internationalization and Localization
-------------------------------------

> Internationalization (`I18n`) is the process of preparing your application to

Reviewer 3: ```suggestion
Internationalization (I18n) is the process of preparing your application to
```

We aren't talking about a constant so it should not be using monospacing font. There are a lot of places where this is happening in this guide

Author: I'll fix this @rafaelfranca, thanks for pointing it out

I18n.enforce_available_locales = true
I18n.exception_handler = MyExceptionHandler.new
I18n.backend = I18n::Backend::Simple.new
```

Internationalizing an Application
---------------------------------

In this section, we'll internationalize a basic Rails application.

### Creating Locale Dictionaries

Rails applications contain a `config/locales` directory with a `en.yml` locale
file by default. This file is a YAML dictionary that contains translations for
the English locale.

The default `en.yml` locale in this directory contains a sample translation:

```yaml
en:
  hello: "Hello world"
```

Looking up the translation for `hello` in the `:en` locale returns `"Hello world"`:

```ruby
I18n.t("hello")
# => "Hello world"
```

The I18n library uses English as the default locale, so when a locale is not
explicitly set, `:en` will be used for looking up translations. To learn how to
change the current locale, see [Managing the Locale across Requests](#managing-the-locale-across-requests).

### Organization of Locale Files

When using the default `Simple` backend shipped with the i18n library,
dictionaries are stored in plain-text files on disk. Putting translations for

> all parts of your application in one file per locale can become hard to manage,

Reviewer 2: TIL you could do that. I've never seen this in the wild. Should we put more emphasis on doing this when things get too big, or that this is not a prescriptive hierarchy. I liked how the previous text mention that users should pick a hierarchy that makes sense for them.

|-----en.yml
|---users
|-----es.yml
|-----en.yml
|---navigation
|-----es.yml
|-----en.yml
```

This way, the model and model attribute names, text inside views, and global
defaults such as date and time formats are all kept separate. Other stores for
the i18n library could provide different means of separation.

### Using Translations in Views and Controllers

Given the following example, we have a `HomeController` with an `index` action
that sets a flash message and renders the `app/views/home/index.html.erb`
template:

```ruby
# config/routes.rb
Rails.application.routes.draw do
  root to: "home#index"
end
```

```ruby
# app/controllers/home_controller.rb
class HomeController < ApplicationController
  def index
    flash[:notice] = "Hello Flash"
  end
end
```

```html+erb
<!-- app/views/home/index.html.erb -->
<h1>Hello World</h1>
<p><%= flash[:notice] %></p>
```

To internationalize this code, replace the strings with calls to Rails' `#t` helper using translation keys:

```ruby
# app/controllers/home_controller.rb
class HomeController < ApplicationController
  def index
    flash[:notice] = t("hello_flash")
  end
end
```

```html+erb
<!-- app/views/home/index.html.erb -->

> <h1><%= t "hello_world" %></h1>

Reviewer 2: Any particular reason to skip parenthesis here? I noticed we used it in the controller and a few other view samples.

```suggestion
<h1><%= t("hello_world") %></h1>
```

<p><%= flash[:notice] %></p>
```

Now, when this view is rendered, it will show an error message indicating that
the translations for the keys `"hello_world"` and `"hello_flash"` are missing.

```
en, hello_world
translation missing: en, hello_flash
```

Now add the translations to your locale files:

```yaml
# config/locales/en.yml
en:
  hello_world: Hello world!
  hello_flash: Hello flash!
```

```yaml
# config/locales/es.yml
es:
  hello_world: Hola mundo!

>   hello_flash: Hola flash!

Reviewer 2: Not a Spanish speaker, but pretty sure we need ¡ on the translated versions
```suggestion
  hello_world: ¡Hola mundo!
  hello_flash: ¡Hola flash!
```

translations from other locale files. Rails does not switch locales from a query
parameter automatically; your application must set the locale for the request,
for example with an `around_action` as shown in
[Managing the Locale across Requests](#managing-the-locale-across-requests).

If the locale is set via the URL to the Spanish locale (`http://localhost:3000?locale=es`), the response renders the Spanish strings:

```
Hola mundo!
Hola flash!
```

NOTE: You need to restart the server when you add new locale files.

### Passing Variables to Translations

One key consideration for successfully internationalizing an application is to
avoid making incorrect assumptions about grammar rules when abstracting
localized code. Grammar rules that seem fundamental in one locale may not hold
true in another.

Improper abstraction is shown in the following example, where assumptions are
made about the ordering of the different parts of the translation.

```erb
<!-- app/views/products/show.html.erb -->
<%= "#{t('currency')}#{@product.price}" %>
```

```yaml
# config/locales/nl.yml
nl:
  currency: "€ "
```

```yaml
# config/locales/es.yml
es:
  currency: "€ "
```

> If the product's price is 100 then Dutch would want `"€ 100"` while Spanish

Reviewer 2: Not to be pedantic, but this isn't true for _all_ Spanish. Just from Spain, I think.

Author: Thanks, I changed this to "might" so that we're not speaking in absolutes.

and currency values.

NOTE: The `default` and `scope` keywords are reserved and can't be used as
variable names. If used, an `I18n::ReservedInterpolationKey` exception is
raised. If a translation expects an interpolation variable, but this has not
been passed to `#translate`, an `I18n::MissingInterpolationArgument` exception
is raised.

### Adding Date/Time Formats

To localize the time format, you can pass the `Time` object to `I18n.l` or,
preferably, use Rails' `#l` helper. You can pick a format by passing the
`:format` option.

```erb
<!-- app/views/home/index.html.erb -->
<h1><%= t "hello_world" %></h1>
<p><%= flash[:notice] %></p>
<p><%= l Time.now, format: :short %></p>
```

For example, you can define a custom short time format in a locale file:

```yaml
# config/locales/es.yml
es:
  time:
    formats:
      short: "%H:%M"
```

Then `l(Time.current, format: :short, locale: :es)` uses that format for the
Spanish locale.

Rails also provides [`relative_time_in_words`](https://api.rubyonrails.org/classes/ActionView/Helpers/DateHelper.html#method-i-relative_time_in_words) for a localized
past/future phrase rather than a formatted timestamp:

```erb

> <p><%= relative_time_in_words(3.minutes.from_now) %></p>

Reviewer 2: Should we put examples of what that renders? Even if as a comment

```erb
<!-- app/views/books/index.html.erb -->
<h1>Books</h1>
```

```erb
<!-- app/views/books/index.es.html.erb -->
<h1>Libros</h1>
```

You can make use of this feature when working with a large amount
of static content, which would be clumsy to put inside YAML or Ruby
dictionaries. Bear in mind, though, that any change you make to the
template must be propagated to all localized variants.

### Inflection Rules for Other Locales

Rails allows you to define inflection rules such as singularization and
pluralization for locales other than English. These rules can be defined for
multiple locales in `config/initializers/inflections.rb`. The initializer
contains a default example for specifying additional rules for English; follow
that format for other locales as you see fit.

Rails already includes default English inflection rules, like the pluralization of "person" to "people":

```ruby
"person".pluralize
# => "people"
```

You can add custom pluralization rules for Spanish, like below:

```ruby
# config/initializers/inflections.rb
ActiveSupport::Inflector.inflections(:es) do |inflect|
  inflect.irregular "el pais", "los paises"

> end

Reviewer 2: I wonder if it is confusing to add inflections to two-word strings. I can't give an example in Spanish, so here's a single-word one in Portuguese
```suggestion
ActiveSupport::Inflector.inflections(:pt_br) do |inflect|
  inflect.irregular "aluguel", "aluguéis"
end
```

**DO NOT READ THIS FILE ON GITHUB, GUIDES ARE PUBLISHED ON <https://guides.rubyonrails.org>.**

Internationalization (I18n)
===========================

> The Ruby [i18n gem](https://github.com/ruby-i18n/i18n), which ships with Ruby on Rails, provides an easy-to-use and extensible framework for translating your application into a single custom language or for providing multi-language support.

Reviewer: Starting with a non-Rails gem in the first sentence is not typical for the guides.
Maybe make the introduction more higher-level and move the link to the opening section?
```suggestion
In this guide you will learn how to translate your application into a single non-English language or provide multi-language support.
```

* How to customize backends and handle translation-related exceptions

--------------------------------------------------------------------------------
Internationalization and Localization
-------------------------------------

Internationalization (I18n) is the process of preparing your application to
support multiple languages and regional formats. In practice, this usually means
abstracting strings and other locale-specific elements, such as date or currency
formats, out of your application.

Localization (L10n) is the process of providing translations and locale-specific formats for those abstracted pieces.

To internationalize a Rails application, you typically need to:

* ensure your application is configured for internationalization (I18n)
* tell Rails where to find translation files
* set, preserve, and switch the locale for each request

To localize your application, you typically need to:

* provide translations for application text
* add or customize locale-specific formats, such as dates, times, and currencies
* organize and maintain translation data

How I18n in Ruby on Rails Works
-------------------------------

> Rails ships with the Ruby i18n gem, which provides the underlying internationalization framework. In this guide, the I18n API refers to the interface your Rails application uses to work with that framework, such as `I18n.t`, `I18n.l`, and Rails translation helpers. As part of this system, static strings in Rails itself, such as Active Record validation messages and time and date formats, are already internationalized.

Reviewer: If the link is removed from the first sentence, the link would be added here.
```suggestion
Rails ships with the Ruby [i18n gem](https://github.com/ruby-i18n/i18n), which provides the underlying internationalization framework. In this guide, the I18n API refers to the interface your Rails application uses to work with that framework, such as `I18n.t`, `I18n.l`, and Rails translation helpers. As part of this system, static strings in Rails itself, such as Active Record validation messages and time and date formats, are already internationalized.
```

|-----en.yml
|---users
|-----es.yml
|-----en.yml
|---navigation
|-----es.yml
|-----en.yml
```

This way, the model and model attribute names, text inside views, and global
defaults such as date and time formats are all kept separate. Other stores for
the i18n library could provide different means of separation.

### Using Translations in Views and Controllers

Given the following example, we have a `HomeController` with an `index` action
that sets a flash message and renders the `app/views/home/index.html.erb`
template:

```ruby
# config/routes.rb
Rails.application.routes.draw do
  root to: "home#index"
end
```

```ruby
# app/controllers/home_controller.rb
class HomeController < ApplicationController
  def index
    flash[:notice] = "Hello Flash"
  end
end
```

```html+erb
<!-- app/views/home/index.html.erb -->
<h1>Hello World</h1>
<p><%= flash[:notice] %></p>
```

To internationalize this code, replace the strings with calls to Rails' `#t` helper using translation keys:

```ruby
# app/controllers/home_controller.rb
class HomeController < ApplicationController
  def index
    flash[:notice] = t("hello_flash")
  end
end
```

```html+erb
<!-- app/views/home/index.html.erb -->

> <h1><%= t("hello_world") %></h1>
> <p><%= flash[:notice] %></p>
> ```
> 
> Now, when this view is rendered, it will show an error message indicating that
> the translations for the keys `"hello_world"` and `"hello_flash"` are missing.
> 
> ```
> en, hello_world
> translation missing: en, hello_flash
> ```

Reviewer: I found this somewhat confusing.
The default en.yml contains `hello: "Hello world"` which is very similar to `hello_world: Hello world!`.
I had to double check `hello_world` was missing.
Also I would expect to see a translation missing for hello_world as well (the warning is present in the HTML though).
Maybe use different examples and show the rendered HTML instead? This would also show where the translation missing warnings can be seen:

```suggestion
Now, when this view is rendered, it will show an error message indicating that
the translations for the keys `"home_title"` and `"home_flash"` are missing.

```html+erb
<h1><span class="translation_missing" title="translation missing: en.home_title">Home Title</span></h1>
<p>Translation missing: en.home_flash</p>
```
```

Now add the translations to your locale files:

```yaml
# config/locales/en.yml
en:
  hello_world: Hello world!
  hello_flash: Hello flash!
```

```yaml
# config/locales/es.yml
es:
  hello_world: ¡Hola mundo!
  hello_flash: ¡Hola flash!
```

With the default locale, the page renders the English strings.

```
Hello world!
Hello flash!
```

Once you start switching locales per request, the same keys can render different
translations from other locale files. Rails does not switch locales from a query
parameter automatically; your application must set the locale for the request,
for example with an `around_action` as shown in
[Managing the Locale across Requests](#managing-the-locale-across-requests).

> If the locale is set via the URL to the Spanish locale (`http://localhost:3000?locale=es`), the response renders the Spanish strings:

Reviewer: Another option would be to skip the complexity of around filters for now and change the default locale config instead.
````suggestion

If the default locale is set to the Spanish locale

```ruby
# config/application.rb
config.i18n.default_locale = :es
```

 the response renders the Spanish strings:
````

<h1>Libros</h1>
```

You can make use of this feature when working with a large amount
of static content, which would be clumsy to put inside YAML or Ruby
dictionaries. Bear in mind, though, that any change you make to the
template must be propagated to all localized variants.

### Inflection Rules for Other Locales

Rails allows you to define inflection rules such as singularization and
pluralization for locales other than English. These rules can be defined for
multiple locales in `config/initializers/inflections.rb`. The initializer
contains a default example for specifying additional rules for English; follow
that format for other locales as you see fit.

Rails already includes default English inflection rules, like the pluralization of "person" to "people":

```ruby
"person".pluralize
# => "people"
```

You can add custom pluralization rules for Portuguese, like below:

```ruby
# config/initializers/inflections.rb
ActiveSupport::Inflector.inflections(:pt_br) do |inflect|
  inflect.irregular "aluguel", "aluguéis"
end
```

Then inflector-based helpers can use those rules for that locale:

```ruby
"aluguel".pluralize(:pt_br)
# => "aluguéis"
```

> When regional locales are used together with `I18n.fallbacks`, inflection rules can also come from a fallback locale. For example, `en-GB` can reuse `:en` inflections when that fallback is mapped explicitly.

Reviewer: Maybe show an example here for a fallback as well (and link to the configuration guide)?
````suggestion
When regional locales are used together with [`I18n.fallbacks`][], inflection rules can also come from a fallback locale. For example, `en-GB` can reuse `:en` inflections when that fallback is mapped explicitly.

```ruby
config.i18n.fallbacks = [:"en-GB": :en]
```

[`I18n.fallbacks`]:
    configuring.html#config-i18n-fallbacks
````

> Rails Internationalization (I18n) API

Author: @p8 I'm thinking of renaming this to Rails Internationalization (I18n) or Internationalization (I18n), what do you think?

Reviewer: Yes, I think that would be an improvement. Everything else also has an API, so it seems redundant.
"Internationalization (I18n)" would have my slight preference.

Also maybe we can move this guide a bit further down in the "Digging Deeper" menu?
Maybe move it under "Working with Javascript"? Readers would probably introduce JavaScript before thinking about I18n?

Author: Done ✅ Thanks @p8

#   config.hosts << 'it.application.local:3000'
def extract_locale_from_subdomain
  parsed_locale = request.subdomains.first
  I18n.available_locales.map(&:to_s).include?(parsed_locale) ? parsed_locale : nil
end
```

If your application includes a locale switching menu, you would then have
something like this in it:

```ruby
link_to("Deutsch", "#{APP_CONFIG[:deutsch_website_url]}#{request.env['PATH_INFO']}")
```

assuming you would set `APP_CONFIG[:deutsch_website_url]` to some value like
`http://www.application.de`.

This solution has the advantages mentioned above, however, you may not be able to
or may not want to provide different localizations ("language versions") on
different domains. The most obvious solution would be to include locale code in
the URL params (or request path).

NOTE: Controller-based locale extraction works well for simple applications. For
production applications that derive the locale from host or path information,
Rack middleware can extract the locale before routing and store it in the Rack
environment. See [Rails on Rack](rails_on_rack.html) for more about Rack
middleware in Rails applications.

> ### Setting the Locale from URL Params

Reviewer 2: I wonder if this should be a subsection the te Request section, or at least show up right after it, since they're tightly linked.

`routes.rb` doesn't take locale into account.

You would need to map URLs like these:

```ruby
# config/routes.rb
get "/:locale" => "dashboard#index"
```

WARNING: Be mindful of the order of your routes so that route declarations do
not unintentionally match other routes.

For applications with more complex path-based locale requirements, locale extraction can be handled at the Rack middleware layer before the request reaches Rails routing. This keeps locale handling separate from the routing layer and is typically only needed when rewriting request paths or mounting the application under a locale-specific prefix.

### Setting the Locale from User Preferences

An application with authenticated users may allow users to set a locale
preference through the application's interface. With this approach, a user's
selected locale is persisted in the database and used to set the locale for
authenticated requests by that user.

When using a stored preference, ensure that the value is included in
`I18n.available_locales` before applying it.

```ruby
# app/controllers/concerns/locale.rb
around_action :switch_locale

def switch_locale(&action)

>   locale = current_user&.locale

Reviewer 2: Nit: should we use CurrentAttributes here? 
```suggestion
  locale = Current.user&.locale
```

end
```

This example uses a URL query parameter to set the locale, such as
`http://example.com/books?locale=pt`. With this approach,
`http://localhost:3000?locale=pt` renders the Portuguese localization, while
`http://localhost:3000?locale=de` loads a German localization.

### Setting the Locale from URL Params

The [previous section](#setting-the-locale-from-a-request-parameter) showed how
to read the locale from `params[:locale]`. In practice, that is only half of the
solution. If the locale comes from the URL, your application also needs to
include it in generated URLs so the choice is preserved from one request to the
next.

For example, you may want URLs such as `www.example.com/books?locale=ja` or
`www.example.com/ja/books`.

This approach has many of the same advantages as using the domain name for the
locale; the locale is visible in the URL, the URL can be shared, and search
engines can distinguish between language-specific pages. It does require a

> little more setup, though.

Reviewer: For the "Locale from Domain Name" section we show the advantages as a list.
I think it would be nice to show the advantages here as a list as well, for easier comparison.
Also this should probably mention it has the same advantages as "using a parameter" (the domain name hasn't been mentioned yet)?
```suggestion
This approach has many of the same advantages as using a parameter for the
locale:
* the locale is visible in the URL
* the URL can be shared
* search engines can distinguish between language-specific pages
 
It does require a little more setup, though.
```

`routes.rb` doesn't take locale into account.

You would need to map URLs like these:

```ruby
# config/routes.rb
get "/:locale" => "dashboard#index"
```

WARNING: Be mindful of the order of your routes so that route declarations do
not unintentionally match other routes.

For applications with more complex path-based locale requirements, locale extraction can be handled at the Rack middleware layer before the request reaches Rails routing. This keeps locale handling separate from the routing layer and is typically only needed when rewriting request paths or mounting the application under a locale-specific prefix.

### Setting the Locale from the Domain Name

You can also set the locale from the domain or subdomain where your application runs.

In order to set the locale from the domain name, you need to have a separate domain for each locale. For example, `www.example.com` can load the English (or the default)
locale, and `www.example.es` the Spanish locale. Setting the locale from
the domain name has several advantages:

* The locale is an obvious part of the URL.
* People intuitively grasp in which language the content will be displayed.

> * It is trivial to implement in Rails.

Reviewer: This probably applies to other solutions as well.
Also you need to setup your application and server to handle multiple domains, which isn't that trivial?
```suggestion
```

language, and IP-based lookups may be affected by VPNs, mobile carriers,
proxies, or corporate networks.

### Storing the Locale from the Session or Cookies

Sessions or cookies can be useful for remembering a user's previous locale
choice, especially when no explicit locale is present in the URL or another
request-based source. In many applications, they are best treated as a fallback
rather than the canonical source of the locale.

The locale is often best kept transparent and reflected in the URL, so that a
shared link shows the same content and language to everyone who opens it. This
also aligns with common RESTful expectations.

Overview of the I18n API Features
---------------------------------

The following sections cover the I18n API in more depth, using examples with
both `I18n.translate` and the [`translate` view helper
method](https://api.rubyonrails.org/classes/ActionView/Helpers/TranslationHelper.html#method-i-translate).

### Looking up Translations

#### Basic Lookup, Scopes, and Nested Keys

Translations are looked up by keys. String keys are a good default because they
work naturally with full keys such as `"books.index.title"` and relative keys

> such as `".title"`:

Reviewer: The term "relative keys" isn't used anywhere else in the guide ("lazy lookup" is the term used later on).
It could be confusing to use a term that is explained later on.
Also, I think "books.index.title" is a good example by itself, ".title" doesn't add that much.

Maybe remove  the mention of relative keys here? 
```suggestion
work naturally with keys such as `"books.index.title"`:
```

```ruby

> I18n.t "message"

Reviewer: Maybe use the "books.index.title" example here as well?
```suggestion
I18n.t "books.index.title"
```

# => "Validation failed: Name can't be blank"
```

This looks up the `"record_invalid"` message in the Active Record error messages.

Additionally, both the key and scopes can be specified as dot-separated keys as in:

```ruby
I18n.translate "activerecord.errors.messages.record_invalid"
# => "Validation failed: Name can't be blank"
```

Thus the following calls are equivalent:

```ruby
I18n.t "activerecord.errors.messages.record_invalid"
# => "Validation failed: Name can't be blank"
I18n.t "errors.messages.record_invalid", scope: "activerecord"
# => "Validation failed: Name can't be blank"
I18n.t "record_invalid", scope: "activerecord.errors.messages"
# => "Validation failed: Name can't be blank"
I18n.t "record_invalid", scope: ["activerecord", "errors", "messages"]
# => "Validation failed: Name can't be blank"
```

#### Defaults

When a `:default` option is given, its value will be returned if the translation is missing:

```ruby
I18n.t "missing", default: "Not here"
# => 'Not here'
```

If the `:default` value is a string key, it will be used as a key and translated. One can provide multiple values as default. The first one that results in a value will be returned.

> The example below first tries to translate the key `"missing"` and then the key `"also_missing"`. Since both do not yield a result, the string "Not here" will be returned:

Reviewer: Nitpick. To me it would be more logical to group the single and array values as separate paragraphs.
```suggestion
If the `:default` value is a string key, it will be used as a key and translated.

One can also provide multiple values as default. The first one that results in a value will be returned.
The example below first tries to translate the key `"missing"` and then the key `"also_missing"`. Since both do not yield a result, the string "Not here" will be returned:
```

### Pluralization

> In many languages,including English, there are only two forms, a singular and a plural, for

Reviewer: ```suggestion
In many languages, including English, there are only two forms, a singular and a plural, for
```

#### Locale-specific Rules

The I18n gem provides a `Pluralization` backend that can be used to enable
locale-specific rules. Include it to the `Simple` backend, then add the
localized pluralization algorithms to translation store, as `i18n.plural.rule`.

```ruby
I18n::Backend::Simple.include(I18n::Backend::Pluralization)
I18n.backend.store_translations :pt, i18n: { plural: { rule: lambda { |n| [0, 1].include?(n) ? :one : :other } } }

> I18n.backend.store_translations :pt, apples: { one: "one or none", other: "more than one" }

Reviewer: Using some linebreaks here makes sure the example doesn't require scrolling, similar to the inbox example above.
```suggestion
I18n.backend.store_translations :pt, i18n: {
  plural: { rule: lambda { |n| [0, 1].include?(n) ? :one : :other } } 
}
I18n.backend.store_translations :pt, apples: {
  one: "one or none",
  other: "more than one"
}
```

The locale can be either set pseudo-globally to `I18n.locale` (which uses `Thread.current` in the same way as, for example, `Time.zone`) or can be passed as an option to `#translate` and `#localize`.

WARNING: The i18n gem stores `I18n.locale` using `Thread.current`. Applications serving multiple requests from the same thread should set the locale explicitly around each request, for example with `I18n.with_locale`.

If no locale is passed, `I18n.locale` is used:

```ruby
I18n.locale = :de
I18n.t "foo"

> I18n.l Time.now

Reviewer: Maybe show what key is used for lookup here?
```suggestion
I18n.t "foo"    # Looks up de.foo
I18n.l Time.now # Looks up de.time.formats.default
```

```

Explicitly passing a locale:

```ruby
I18n.t "foo", locale: :de

> I18n.l Time.now, locale: :de

Reviewer: Maybe use different locales here to make the examples clearer?
```suggestion
I18n.t "foo", locale: :br    # Looks up br.foo
I18n.l Time.now, locale: :es # Looks up es.time.formats.default
```

> <div><%= t('title.html') %></div>

Reviewer: The red HTML tags really stand out:
<img width="860" height="388" alt="Image" src="https://github.com/user-attachments/assets/8827132d-4814-4b22-864c-8e321b177874" />

What do you think about removing the extra divs here to make it clearer what HTML tags are rendered?
This would also mean we should probably delete the screenshot.
```suggestion
<%= t('welcome') %>
<%= raw t('welcome') %>
<%= t('hello_html') %>
<%= t('title.html') %>
```

```

This renders:

```html
<div>&lt;b&gt;welcome!&lt;/b&gt;</div>
<div><b>welcome!</b></div>
<div><b>hello!</b></div>

> <div><b>title!</b></div>

Reviewer: ```suggestion
&lt;b&gt;welcome!&lt;/b&gt;
<b>welcome!</b>
<b>hello!</b>
<b>title!</b>
```

```erb
<%= form_with model: @user do |form| %>
  <%= form.label :login %>
  <%= form.text_field :login %>
<% end %>
```

With the following translations:

```yaml
en:
  activerecord:
    attributes:
      user:
        login: "Handle"
```

the label renders as:

```html
<label for="user_login">Handle</label>
```

### Other I18n Helpers

Rails uses fixed strings and other localizations, such as format strings and other format information in a couple of helpers. Here's a brief overview.

#### Action View Helper Methods

* [`distance_of_time_in_words`](https://api.rubyonrails.org/classes/ActionView/Helpers/DateHelper.html#method-i-distance_of_time_in_words)
  translates and pluralizes its result and interpolates the number of seconds,
  minutes, hours, and so on.

    ```ruby
    distance_of_time_in_words(Time.current, Time.current + 3.minutes)
    # => "3 minutes"
    ```

    It uses translations from the `datetime.distance_in_words` scope.

* [`relative_time_in_words`](https://api.rubyonrails.org/classes/ActionView/Helpers/DateHelper.html#method-i-relative_time_in_words)
  builds on `distance_of_time_in_words` and adds a localized past/future

>   wrapper.

Reviewer: I found the API documentation description a bit more clear.
Maybe:
```suggestion
  builds on `distance_of_time_in_words` and adds a localized prefix/suffix
  depending on whether the time is in the past or future.
```

    ```ruby
    relative_time_in_words(3.minutes.from_now)

>     # => "in 3 minutes"

Reviewer: Might as well add the past case here as well:
```suggestion
    # => "in 3 minutes"
    relative_time_in_words(3.minutes.ago)
    # => "3 minutes ago"
```

* [`ActiveModel::Error#full_message`](https://api.rubyonrails.org/classes/ActiveModel/Error.html#method-i-full_message)
  and [`ActiveModel::Errors#full_messages`](https://api.rubyonrails.org/classes/ActiveModel/Errors.html#method-i-full_messages)
  prepend the attribute name to the error message using a format looked up from
  `errors.format` (default: `"%{attribute} %{message}"`).

    ```ruby
    user.errors.full_messages
    # => ["Name can't be blank"]
    ```

    To customize the default format, override it in the app's locale files. To customize the format per model or per attribute, see [`config.active_model.i18n_customize_full_message`][].

[`config.active_model.i18n_customize_full_message`]: configuring.html#config-active-model-i18n-customize-full-message

#### Active Support Methods

* [`Array#to_sentence`](https://api.rubyonrails.org/classes/Array.html#method-i-to_sentence) uses format settings from the `support.array` scope.

    ```ruby
    ["apples", "oranges", "pears"].to_sentence
    # => "apples, oranges, and pears"
    ```

Advanced Configuration and Setup
---------------------------------

> ### Configure the load paths and default locale

Reviewer: Nitpick, most other headings use the "present participle" verb form (setting, adding, passing):
```suggestion
### Configuring the load paths and default locale
```

Rails automatically adds translation files from `config/locales` to the
translations load path. By default, this includes `.rb` and `.yml` files, and
for many applications that is all you need.

Rails' own translations are organized the same way. See, for example, the
Active Model validation messages in
[`activemodel/lib/active_model/locale/en.yml`](https://github.com/rails/rails/blob/main/activemodel/lib/active_model/locale/en.yml)
or the date and time formats in
[`activesupport/lib/active_support/locale/en.yml`](https://github.com/rails/rails/blob/main/activesupport/lib/active_support/locale/en.yml).
With the default `Simple` backend, you can store translations in YAML files or in
plain Ruby hashes.

> NOTE: The i18n library takes a pragmatic approach to locale keys, including only the _locale_ ("language") part, like `:en`, `:pl`, not the _region_ part, like `:"en-US"` or `:"en-GB"`, which are traditionally used for separating "languages" and "regional setting" or "dialects". Many international applications use only the "language" element of a locale such as `:cs`, `:th`, or `:es` (for Czech, Thai, and Spanish). However, there are also regional differences within different language groups that may be important. For instance, in the `:"en-US"` locale you would have $ as a currency symbol, while in `:"en-GB"`, you would have £. Nothing stops you from separating regional and other settings in this way: you just have to provide full "English - United Kingdom" locale in a `:"en-GB"` dictionary.

Reviewer: I'm not sure I understand this NOTE.
`rails`, `i18n` and `i18n-rails` all support the region part in locales.
Is the pragmatic approach that excluding the region is supported as well?

Maybe it becomes clearer if we remove the first sentence?
Also, this NOTE seems more at home after we mention "The I18n library uses English as the default locale", as it describes why we use `:en` and not `:"en-US"`.

```suggestion
NOTE: Many international applications use only the "language" element of a locale such as `:cs`, `:th`, or `:es` (for Czech, Thai, and Spanish). However, there are also regional differences within different language groups that may be important. For instance, in the `:"en-US"` locale you would have $ as a currency symbol, while in `:"en-GB"`, you would have £. Nothing stops you from separating regional and other settings in this way: you just have to provide full "English - United Kingdom" locale in a `:"en-GB"` dictionary.
```

```

In general, prefer `config.i18n.*` in application configuration. Appending
directly to `I18n.load_path` will _not_ override translations from external
gems in the same way.

You can use YAML (`.yml`) or plain Ruby (`.rb`) files to store translations in
the `Simple` backend. YAML is the format most Rails applications use, but it is
sensitive to whitespace and special characters, so syntax mistakes can prevent a
locale file from loading correctly.

Ruby locale files can be useful when you want syntax errors to fail fast, or
when you need values such as lambdas in your locale data.

If your translations are stored in YAML files, certain keys must be escaped. They are:

* true, on, yes
* false, off, no

Examples:

```yaml
# config/locales/en.yml
en:
  success:
    'true':  'True!'
    'on':    'On!'
    'false': 'False!'
  failure:
    true:    'True!'
    off:     'Off!'
    false:   'False!'
```

```ruby
I18n.t "success.true"  # => 'True!'
I18n.t "success.on"    # => 'On!'
I18n.t "success.false" # => 'False!'
I18n.t "failure.false" # => Translation Missing
I18n.t "failure.off"   # => Translation Missing
I18n.t "failure.true"  # => Translation Missing
```

> ### Store Custom Translations

Reviewer: Nitpick, most other headings use the "present participle" verb form:
```suggestion
### Storing Custom Translations
```

### Using Alternate Backends

For several reasons the `Simple` backend shipped with Active Support only does the
"simplest thing that could possibly work" for Ruby on Rails. It is guaranteed
to work for English and, as a side effect, for languages that are very similar
to English. It is also only capable of reading translations and cannot
dynamically store them in another format.

That does not mean you're stuck with these limitations, though. The Ruby I18n gem makes it very easy to exchange the `Simple` backend implementation with something else that fits better for your needs, by passing a backend instance to the `I18n.backend=` setter.

For example, you can replace the `Simple` backend with the `Chain` backend to chain multiple backends together. This is useful when you want to use standard translations with a `Simple` backend but store custom application translations in a database or other backends.

With the `Chain` backend, you could use the Active Record backend and fall back to the (default) `Simple` backend:

```ruby

> I18n.backend = I18n::Backend::Chain.new(I18n::Backend::ActiveRecord.new, I18n.backend)

Reviewer: This example currently requires scrolling:

<img width="910" height="106" alt="Image" src="https://github.com/user-attachments/assets/ad0df54e-2680-4fd5-b13f-bb3c8d15a261" />

```suggestion
I18n.backend = I18n::Backend::Chain.new(
  I18n::Backend::ActiveRecord.new,
  I18n.backend
)
```

```

In this configuration, lookups first check the Active Record backend and then fall back to the locale files loaded by the `Simple` backend:

```ruby
I18n.t("welcome")

> # Looks for "welcome" in the Active Record backend first, then in YAML/Ruby locale files.

Reviewer: This one isn't fully readable because of the copy button:

<img width="832" height="95" alt="Image" src="https://github.com/user-attachments/assets/9b3632e3-80a2-4484-9fa7-c74fca819ae3" />

```suggestion
# Looks for "welcome" in the Active Record backend first,
# then in YAML/Ruby locale files.
```

| `I18n::MissingTranslationData`       | no translation was found for the requested key                                                    |
| `I18n::InvalidLocale`                | the locale set to `I18n.locale` is invalid (e.g. `nil`)                                           |
| `I18n::InvalidPluralizationData`     | a count option was passed but the translation data is not suitable for pluralization              |
| `I18n::MissingInterpolationArgument` | the translation expects an interpolation argument that has not been passed                        |
| `I18n::ReservedInterpolationKey`     | the translation contains a reserved interpolation variable name (i.e. one of: `scope`, `default`) |
| `I18n::UnknownFileType`              | the backend does not know how to handle a file type that was added to `I18n.load_path`            |

#### Handling Missing Translations

If `config.i18n.raise_on_missing_translations` is `true`, missing translations
raise errors instead of returning the usual `"Translation missing: ..."`
message. If the value is `:strict`, model translations also raise. It is a good
idea to enable this in your test environment so you can catch missing

> translations early.

Reviewer: Maybe swap this with the "default `config.i18n.raise_on_missing_translations` is `false`" paragraph?
So we first show the Rails default, and then we expand with the example of setting it to `true`.

> 

Reviewer: Maybe add a raising example here as well?
````suggestion

For example:

```ruby
I18n.t("missing")
=> Translation missing: en.missing (I18n::MissingTranslationData)
```

````
